mlflow/server/js/src/components/MetricsPlotPanel.js

```diff
@@ -36,23 +36,27 @@ export class MetricsPlotPanel extends React.Component {
     runDisplayNames: PropTypes.arrayOf(String).isRequired,
   };
 
+  // The fields below are exposed as instance attributes rather than component state so that they
+  // can be updated without triggering a rerender.
+  //
+  // ID of Javascript future (created via setTimeout()) used to trigger legend-click events after a
+  // delay, to allow time for double-click events to occur
+  legendClickTimeout = null;
+  // Time (millis after Unix epoch) since last legend click - if two clicks occur in short
+  // succession, we trigger a double-click event & cancel the pending single-click.
+  prevLegendClickTime = Math.inf;
+
   constructor(props) {
     super(props);
-    const plotMetricKeys = Utils.getPlotMetricKeysFromUrl(props.location.search);
-    const selectedMetricKeys = plotMetricKeys.length ? plotMetricKeys : [props.metricKey];
     this.state = {
-      selectedXAxis: X_AXIS_RELATIVE,
-      selectedMetricKeys,
-      showPoint: false,
       historyRequestIds: [],
-      lineSmoothness: 0,
-      // Used to keep track of the most-recent linear y-axis plot range, to handle the specific
-      // case where we toggle a plot with negative y-axis bounds from linear to log scale,
-      // and then back to linear scale (we save the initial negative linear y-axis bounds so
-      // that we can restore them when converting from log back to linear scale)
-      lastLinearYAxisRange: [],
     };
-    this.loadMetricHistory(this.props.runUuids, this.state.selectedMetricKeys);
+    const state = this.getUrlState();
+    this.loadMetricHistory(this.props.runUuids, state.selectedMetricKeys);
+  }
+
+  getUrlState() {
+    return Utils.getMetricPlotStateFromUrl(this.props.location.search);
```

Comment:
Good question, AFAICT this works in the same way as MetricsView used to work, where updates to the page URL (`history.push`) cause a rerender of the entire component tree (but React intelligently reuses existing DOM elements where possible to avoid flashing large portions of the page). I experimentally verified this is the case by adding a `console.log` in MetricPage's `render` method, and observed `MetricPage.render` being called both when UI actions update the URL & when manually editing the URL myself